### PR TITLE
[TEAM 3] Agent V2 PR

### PR DESCRIPTION
**Pull request by Neelesh99**
# Summary

Team 3 presents Agrim 2.
A lot of fine-tuning of algorithms. Better foraging, taxing and allocation. Also, some fun stuff to play with.

## Additional Information

Main fixes:

- Foraging doesn't bet everything we have on deer anymore
- When we are president we don't allocate everything to us anymore
- IITO participation has been tuned
- Our disaster predictions have more reasonable confidence levels
- Rule Shifting added
- Compliance tuned
- adv package added with a couple of fun tools (feel free to use)

## Test Plan

- Most things are tested ;)

### CODE REVIEW

File: internal/clients/team3/adv/Malice.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
+package adv
```
Comment (lhl2617) [inline internal/clients/team3/adv/Malice.go]: files should be `lowercase.go` but ok

Comment (Neelesh99) [reply internal/clients/team3/adv/Malice.go]: Fair point, changed in latest commit

File: internal/clients/team3/clientgeneral.go [hunk 2]
```diff
@@ -356,13 +338,3 @@ func (c *client) ResourceReport() shared.ResourcesReport {
 	skewedResource := resource / shared.Resources(c.params.resourcesSkew)
```
Comment (NGovani) [inline internal/clients/team3/clientgeneral.go]: can this ever be 0?

File: internal/clients/team3/clientiitoiifo.go [hunk 3]
```diff
@@ -138,11 +118,11 @@ func (c *client) ReceiveDisasterPredictions(receivedPredictions shared.ReceivedD
 
 	// Add other island's predictions using their confidence values
 	for islandID, prediction := range receivedPredictions {
-		totalCoordinateX += c.trustScore[islandID] * prediction.PredictionMade.Confidence * prediction.PredictionMade.CoordinateX
-		totalCoordinateY += c.trustScore[islandID] * prediction.PredictionMade.Confidence * prediction.PredictionMade.CoordinateY
-		totalMagnitude += c.trustScore[islandID] * prediction.PredictionMade.Confidence * prediction.PredictionMade.Magnitude
-		totalTimeLeft += uint(math.Round(c.trustScore[islandID]*prediction.PredictionMade.Confidence)) * prediction.PredictionMade.TimeLeft
-		totalConfidence += c.trustScore[islandID] * prediction.PredictionMade.Confidence
+		totalCoordinateX += c.trustScore[islandID] / 100 * prediction.PredictionMade.Confidence * prediction.PredictionMade.CoordinateX
```
Comment (NGovani) [inline internal/clients/team3/clientiitoiifo.go]: div by 0 checks

File: internal/clients/team3/president.go [hunk 4]
```diff
@@ -127,25 +150,33 @@ func (p *president) EvaluateAllocationRequests(resourceRequest map[shared.Client
...
-				resourcesRequired = disaster.Magnitude - float64(gameState.CommonPool)/float64(disaster.TimeLeft)
-			}
-		}
+	// Aim to have 100 in common pool after iigo run
+	resourcesRequired := math.Max((float64(p.c.getIIGOCost())+100.0)-float64(gameState.CommonPool), 0)
+	p.c.clientPrint("Resources required in common pool %f", resourcesRequired)
+
+	if len(p.c.globalDisasterPredictions) > int(p.c.ServerReadHandle.GetGameState().Turn) {
+		disaster := p.c.globalDisasterPredictions[int(p.c.ServerReadHandle.GetGameState().Turn)]
+		resourcesRequired += disaster.Magnitude - float64(gameState.CommonPool)/float64(disaster.TimeLeft+1)
```
Comment (NGovani) [inline internal/clients/team3/president.go]: can disaster.TimeLeft ever be -1

**Review by NGovani — CHANGES_REQUESTED** [review]
some div by 0 issues

File: internal/clients/team3/clientgeneral.go (hunk 2 above)
Comment (Neelesh99) [reply internal/clients/team3/clientgeneral.go]: I'm not sure, div 0 check has been added

File: internal/clients/team3/clientiitoiifo.go (hunk 3 above)
Comment (Neelesh99) [reply internal/clients/team3/clientiitoiifo.go]: Yup added for all now, thanks

File: internal/clients/team3/president.go (hunk 4 above)
Comment (Neelesh99) [reply internal/clients/team3/president.go]: No idea really, I've added the checks now

**Review by NGovani — APPROVED** [review]
lets get it in

**Pull request merged**